crates/hyperqueue/src/stream/reader/streamdir.rs

```diff
@@ -0,0 +1,464 @@
+use crate::common::error::HqError;
+use crate::server::event::bincode_config;
+use crate::transfer::stream::{ChannelId, StreamChunkHeader};
+use crate::worker::streamer::{StreamFileHeader, STREAM_FILE_HEADER};
+use crate::{JobId, JobTaskId, Set};
+use bincode::Options;
+use chrono::{DateTime, Utc};
+use colored::{Color, Colorize};
+use lru::LruCache;
+use serde::{Deserialize, Serialize};
+use serde_json::json;
+use smallvec::SmallVec;
+use std::collections::BTreeMap;
+use std::fs::File;
+use std::io::{BufReader, BufWriter, Read, Seek, Write};
+use std::num::NonZeroUsize;
+use std::ops::Deref;
+use std::path::{Path, PathBuf};
+use tako::InstanceId;
+
+#[derive(Clone)]
+pub struct ChunkInfo {
+    time: DateTime<Utc>,
+    position: u64,
+    size: u32, // Currently chunk is actually limited to 128kB
+}
+
+pub struct InstanceInfo {
+    instance_id: InstanceId,
+    channels: [Vec<ChunkInfo>; 2],
+    file_idx: usize,
+    finished: bool,
+}
+
+#[derive(Default)]
+pub struct TaskInfo {
+    instances: SmallVec<[InstanceInfo; 1]>,
+}
+
+impl InstanceInfo {
+    fn channel_size(&self, channel_id: ChannelId) -> u64 {
+        self.channels[channel_id as usize]
+            .iter()
+            .map(|x| x.size as u64)
+            .sum()
+    }
+}
+
+impl TaskInfo {
+    pub fn last_instance(&self) -> &InstanceInfo {
+        self.instances.last().unwrap()
+    }
+
+    pub fn superseded(&self) -> impl Iterator<Item = &InstanceInfo> {
+        self.instances[0..self.instances.len() - 1].iter()
+    }
+
+    pub fn instance_mut(&mut self, instance_id: InstanceId) -> Option<&mut InstanceInfo> {
+        self.instances
+            .iter_mut()
+            .find(|info| info.instance_id == instance_id)
+    }
+
+    pub fn instance(&self, instance_id: InstanceId) -> Option<&InstanceInfo> {
+        self.instances
+            .iter()
+            .find(|info| info.instance_id == instance_id)
+    }
+}
+
+pub struct StreamDir {
+    paths: Vec<PathBuf>,
+    index: BTreeMap<JobId, BTreeMap<JobTaskId, TaskInfo>>,
+    cache: LruCache<usize, BufReader<File>>,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct Summary {
+    pub n_files: u32,
+    pub n_jobs: u32,
+    pub n_tasks: u64,
+    pub n_streams: u64,
+    pub n_opened: u64,
+    pub stdout_size: u64,
+    pub stderr_size: u64,
+    pub n_superseded: u64,
+    pub superseded_stdout_size: u64,
+    pub superseded_stderr_size: u64,
+}
+
+impl StreamDir {
+    pub fn open(path: &Path, server_uid: Option<&str>) -> crate::Result<Self> {
+        log::debug!("Reading stream dir {}", path.display());
+        let mut paths = Vec::new();
+        let mut server_uids: Set<String> = Set::new();
+        let mut found = false;
+        for entry in std::fs::read_dir(path)? {
+            let entry = entry?;
+            let path = entry.path();
+            if path
+                .extension()
+                .and_then(|e| e.to_str())
+                .map(|s| s == "hqs")
```

Comment:
Could we please put `"hqs"` into some global constant? It is probably used on multiple places.